[CIR] Initial implementation of lowering CIR to MLIR

andykaylor opened the pull request:
Add support for lowering CIR to more general MLIR and emitting an MLIR text file, including a new command-line option for emitting MLIR via ClangIR. This does not yet support lowering to LLVM IR or other target-specific formats, but that is the ultimate goal, which is why the newly added directory is named "ThroughMLIR".

The intent here is to allow MLIR optimizations that don't recognize the ClangIR dialect to be run before LLVM IR is generated.

Lowering of global pointers to MLIR is not yet supported (here or in the incubator). A test is added here to track that.

andykaylor committed:
[CIR] Initial implementation of lowering CIR to MLIR

Add support for lowering CIR to MLIR and emitting an MLIR text file.

Lowering of global pointers is not yet supported.

erichkeane commented on clang/include/clang/CIR/CIRGenerator.h:
```diff
@@ -55,6 +55,7 @@ class CIRGenerator : public clang::ASTConsumer {
   void Initialize(clang::ASTContext &astContext) override;
   bool HandleTopLevelDecl(clang::DeclGroupRef group) override;
   mlir::ModuleOp getModule() const;
+  mlir::MLIRContext &getContext() { return *mlirContext; }
```
```suggestion
  mlir::MLIRContext &getMLIRContext() { return *mlirContext; }
```

MINOR preference?  We have all sorts of `contexts` all over the place that these are incredibly confusing.

andykaylor replied:
That's reasonable. I frequently find myself wondering which context things are returning (recently it's been the AST context coming up during IR generation).

dkolsen-pgi replied:
A little while back I changed all the variable, parameter, and field names in the ClangIR code to be `astContext` or `mlirContext`, getting rid of all generic `context` or `cxt` names.  I didn't do the same for function names.  But it would be good if those were also changed.

andykaylor commented on clang/include/clang/Driver/Options.td:
```diff
@@ -2958,6 +2958,8 @@ defm clangir : BoolFOption<"clangir",
   BothFlags<[], [ClangOption, CC1Option], "">>;
 def emit_cir : Flag<["-"], "emit-cir">, Visibility<[ClangOption, CC1Option]>,
   Group<Action_Group>, HelpText<"Build ASTs and then lower to ClangIR">;
+def emit_cir_mlir : Flag<["-"], "emit-cir-mlir">, Visibility<[CC1Option]>, Group<Action_Group>,
```
Agreed. One problem with aligning the two is that in flang, currently, -emit-mlir means to emit the fir dialect, whereas here we wanted it to mean, emit MLIR but **not** the cir dialect. This highlights the problem that "emit MLIR" isn't very specific. Even the interpretation used in the ClangIR incubator is emitting a handful of different dialects (affine, scf, arith, memref, vector, etc.).

clementval replied:
Yeah we have currently two options. `-emit-fir` and ` -emit-hlfir`. These are dump the IR at different level. But as Andy mentioned, it is with the fir or hlfir dialect.

erichkeane replied:
Is there a perhaps a term that MEANS 'generic dialects' or something that we could use here instead? 

perhaps `emit-std-mlir` ? And point out th at this is intending to lower to the standard dialects?

andykaylor replied:
I think "standard" dialects is somewhat contradictory to the MLIR philosophy, though in the abstract such a set seems to exist. There's just no good way to define it as far as I can see. I want to say this lowering would be to language-independent dialects, but I'm not sure where that leaves something like the OpenACC dialect.

andykaylor replied:
I guess 'core' is a good term. Even on that link, "core" and "contributed" dialects are mixed without designation of which is which, and I expect that will grow over time, but it's probably fine to say we're lowering to "core" dialects even if it's an arbitrary subset of what other people might think of as core dialects.

How about `-emit-core-mlir`?

erichkeane replied:
works for me.  And yeah, I think that is what I meant instead of 'dialects' above.

xlauko replied:
This is being reworked in incubator: https://github.com/llvm/clangir/pull/1316

I agree that `-emit-mlir` is ambiguous, though @bcardosolopes wanted to keep original semantics there for now.
TLDR we are working towards `-emit-mlir=<name-of-target-dialect-set>`, e.g., `-emit-mlir=std` (can be renamed to core), `-emit-mlir=cir`, `-emit-mlir=llvm`.

Where these need to cooperate with `-f[no]-clangir-direct-lowering` to translate through `core` dialects or directly from `cir` to `llvm dialect`.

xlauko replied:
This might unify with flang using `-emit-mlir=fir` or `hlfir`.

bcardosolopes replied:
I wasn't aware of the conflict with `emit-mlir` for the flag, bummer. That changes how we should do things in the incubator and make it easier for upstreaming. If we add the `_EQ` version of the flag I guess it's not hard to redirect the flat `-emit-mlir` to expand to `-emit-mlir=<...>` with the proper FIR / CIR or MLIR dialect style in question.

clementval replied:
We don't have a lowering to core dialects at the moment. At least two attempts have been made for that but nothing is upstream so far.

clementval replied:
I double checked in flang and we have a couple of tests (7 files) using that option but we can switch these to use `-emit-fir` or -`emit-hlfir`. 

@banach-space Since you are probably the initial author of this option, do you have any objection to make `-emit-mlir` identical for clang and flang?

andykaylor replied:
> If we add `emit_mlir_EQ` will that peacefully coexist with the `def emit_mlir : Flag<["-"], "emit-mlir">, Alias<emit_fir>;` definition until we can update flang to align the behavior of the option with no argument?

Answering my own question, yes, it will peacefully coexist. At least, the clang part of it worked. I can't see a reason that flang would be impacted (functionally) by what I did here, but I didn't build and test flang.

I think it makes sense to put `emit_mlir_EQ` in here and then have a separate PR to have flang make use of this option, adding the additional flang dialects, and flang-specific help text.

clementval replied:
> I think it makes sense to put emit_mlir_EQ in here and then have a separate PR to have flang make use of this option, adding the additional flang dialects, and flang-specific help text.

Yeah we can address that in a follow up PR if it doesn't break flang with this one.

andykaylor committed:
Change cl option to -emit-core-mlir, change getContext to getMLIRContext()

AaronBallman commented on clang/lib/CIR/FrontendAction/CIRGenAction.cpp:
```diff
@@ -90,6 +92,15 @@ class CIRGenConsumer : public clang::ASTConsumer {
       }
       break;
+    case CIRGenAction::OutputType::EmitMLIR: {
+      auto LoweredMlirModule = lowerFromCIRToMLIR(MlirModule, MlirCtx);
```
Please spell out the type.

AaronBallman commented on clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRToMLIR.cpp:
```diff
@@ -0,0 +1,201 @@
+
+    mlir::OpBuilder b(moduleOp.getContext());
+
+    const auto cirSymType = op.getSymType();
+    assert(!cir::MissingFeatures::convertTypeForMemory());
+    auto convertedType = getTypeConverter()->convertType(cirSymType);
```
Please spell out the types...

AaronBallman commented on clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRToMLIR.cpp:
```diff
@@ -0,0 +1,201 @@
+    auto convertedType = getTypeConverter()->convertType(cirSymType);
+    if (!convertedType)
+      return mlir::failure();
+    auto memrefType = dyn_cast<mlir::MemRefType>(convertedType);
```
`auto *`, right?

andykaylor replied:
No. Many MLIR objects are wrappers around pointers with the necessary overloads to make them act like pointers and be used with `dyn_cast`, `isa`, etc., but they aren't pointers. This is such a case.

AaronBallman replied:
Ah, good to know, thank you!

AaronBallman commented on clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRToMLIR.cpp:
```diff
@@ -0,0 +1,201 @@
+
+  pm.addPass(createConvertCIRToMLIRPass());
+
+  auto result = !mlir::failed(pm.run(mlirModule));
```
Please spell out the type.

AaronBallman commented on clang/test/CIR/Lowering/ThroughMLIR/global.cpp:
```diff
@@ -0,0 +1,79 @@
+// CHECK: memref.global "public" @ld : memref<f80>
+
+__float128 f128;
+// CHECK: memref.global "public" @f128 : memref<f128>
```
What about other builtin types like `__fp16`/`half`, `__ibm128`, `bool`, fixed point types? I suppose another interesting type might be `nullptr_t`?

andykaylor replied:
I think half and __fp16 are covered by _Float16 for the purposes of this test. The other types you mentioned aren't supported here yet.

AaronBallman commented on clang/include/clang/CIR/CIRGenerator.h:
```diff
@@ -55,6 +55,7 @@ class CIRGenerator : public clang::ASTConsumer {
   void Initialize(clang::ASTContext &astContext) override;
   bool HandleTopLevelDecl(clang::DeclGroupRef group) override;
   mlir::ModuleOp getModule() const;
+  mlir::MLIRContext &getMLIRContext() { return *mlirContext; }
```
One overall concern I have is that we're adding an entirely new interface to Clang where we could be introducing const-correctness from the start... but we're not doing it. That's not specific to this PR, but it is something I think should be addressed early on. Retrofitting const-correctness is hard and I realize this is touching other interfaces like MLIR and LLVM, but because this is a brand new component, we should be adding const correctness from the start everywhere possible. (IMO, if MLIR lacks const correctness, that's a problem that should be addressed in MLIR given it's also a new interface, comparatively speaking.)

xlauko replied:
MLIR design is to literally never use const on its primitives: https://mlir.llvm.org/docs/Rationale/UsageOfConst/ This will probably fight back trying to be const correct only on clang side.

AaronBallman replied:
Wow, that's disappointing to say the least.

andykaylor replied:
OK, so to stay consistent with the MLIR design philosophy, we won't use const with MLIR operations or values, but we should be moving everything else to const-correctness as it is upstreamed. Does that sound reasonable?

AaronBallman replied:
Yes, that sounds like a good path forward. FWIW, I think it's reasonable for CIR to have *some* ugly `const_cast` uses to hide the lack of const correctness from Clang, but not to an obnoxious amount (which it sounds like would be the case with MLIR operations or values). Thanks!

andykaylor committed:
Fix autos, add const getMLIRContext()

andykaylor committed:
Use -emit-mlir=[core|cir] command-line options

bcardosolopes commented on clang/include/clang/Driver/Options.td:
```diff
@@ -2958,8 +2958,13 @@ defm clangir : BoolFOption<"clangir",
+def emit_mlir_EQ : Joined<["-"], "emit-mlir=">, Visibility<[CC1Option]>, Group<Action_Group>,
+  HelpText<"Build ASTs and then generate/lower to the selected MLIR dialect, emit the .mlir or .cir file. "
+  "Allowed values are `core` for MLIR core dialects and `cir` for ClangIR">,
+  Values<"core,cir">,
```
In CIR terms, considering we want to cover DirectToLLVM, ThroughMLIR (core) and CIR, shouldn't this be 3 state: `core,llvm,cir`?

andykaylor replied:
I think eventually we will also want to put cir-flat in here, and maybe other variations, but right now these two are the only ones implemented upstream. Technically, you can also lower to the LLVM dialect, but since that's only just now being implemented in the incubator as an output, it isn't done here yet.

bcardosolopes replied:
Makes sense! Thanks

keryell commented:
Is there any emergency to up-stream this compiler flow compared to just focusing on upstreaming the direct-through-LLVM path first?
My experiments with https://github.com/llvm/clangir/pull/1334 around https://github.com/llvm/clangir/issues/1219 makes me feel that it might be better to spend more time in the incubator.

andykaylor commented:
> Is there any emergency to up-stream this compiler flow compared to just focusing on upstreaming the direct-through-LLVM path first? My experiments with [llvm/clangir#1334](https://github.com/llvm/clangir/pull/1334) around [llvm/clangir#1219](https://github.com/llvm/clangir/issues/1219) makes me feel that it might be better to spend more time in the incubator.

No, there is no urgency. This is just something I started on while waiting for the initial function implementation to land. I'm just trying to open up more possible paths for parallel work in the upstreaming. I agree that lowering to MLIR shouldn't be a priority at this point.

keryell commented:
I appreciate the parallelism for sure! :smile:
It would be nice if your team can book more time to work on the fundamental issues related to CIR→MLIR in the incubator.

bcardosolopes reviewed (approved):
LGTM once remaining comments from others are addressed.

keryell reviewed (approved):
Thank you!

andykaylor commented:
@AaronBallman , @erichkeane I think I've addressed the review comments. Is there anything else you'd like changed?

erichkeane reviewed (approved):
I'm happy with this as is, though note Aaron is away at WG14 in Graz this week.  But it looks like you've done everything he's asked for, so feel free to merge.

andykaylor committed:
Update to realign with incubator changes

andykaylor committed:
Update to fix issues after rebase

andykaylor commented:
I've rebased this PR to bring in recent changes, fixed a few problems that were caused by recent changes, and generally refactored the command-line handling to align with a similar change that was made in the incubator.

This doesn't lower the ops and types that have been added since this PR was originally created. I'll handle that in a follow-up PR.

andykaylor committed:
Fix formatting

andykaylor commented:
We agreed at the clangir upstreaming meeting today that this PR should be put on hold. We don't intend to keep lowering to MLIR standard dialects up-to-date as we upstream the rest of the implementation, and it is currently missing a lot of functionality in the incubator, so it seems better to leave this for a future effort.

keryell commented:
> We agreed at the clangir upstreaming meeting today that this PR should be put on hold. 

There is an up-streaming meeting? :thinking:

andykaylor commented:
> There is an up-streaming meeting? 🤔

Yeah. It's not on the LLVM calendar yet, but that's on my TODO list. We're meeting every two weeks. I'll get it on the calendar before the next one.

andykaylor commented:
I marked this as "draft" to indicate that we don't intend to merge it at this time.
